### DISCUSSION TITLE
Timesheet Functionality

### DISCUSSION THREAD

**Discussion by wunter8** [TriliumNext/Trilium#2528] [Show and tell]
I wanted a way to track the amount of time I spend working on a project. Here's what I came up with.

Each time you add a child note to the timesheet, it will automatically change the title of the new child note to `IN - ` or `OUT - ` followed by the current timestamp. After "clocking in", the timesheet's title will show `ON THE CLOCK`. Once you "clock out", the total amount of time spent will be added to the timesheet's title. These totals can be passed up the tree if you have nested timesheets.

## Features:
- automatically decide if you're clocking in or out
- show when you're "on the clock"
- total the amount of hours you worked each time you clock out
- (optional) sum the total hours worked for nested timesheets

## Example 1
![Screenshot from 2022-01-07 15-03-35](https://user-images.githubusercontent.com/8421688/148613592-2530e7e1-76c6-4058-ac31-0d9460deaf72.png)
- Project 1 has 3 children notes (Task 1, Task 2, Task 3)
- Task 1 doesn't have any sub-tasks and is its own timesheet. Totals are combined into Task 1's title
- Task 2 has 2 sub-tasks (Sub-task 1, Sub-task 2). Each sub-task is its own timesheet. Their totals are combined into Task 2's title
- Task 3 also has 2 sub-tasks (Sub-task 1, Sub-task 2). Each sub-task is its own timesheet. Their totals are _not_ combined into Task 3's title
- I am currently "on the clock" on Task 3, Sub-task 2
- In this current configuration, Project 1 total includes times from Task 1 and Task 2 but not Task 3

## Example 2
![Screenshot from 2022-01-07 15-06-39](https://user-images.githubusercontent.com/8421688/148613606-1b31bbc5-761d-4a54-92e2-69e0b17d55c2.png)
- This shows another possible configuration with deeper nesting using a Year, Month, Day, Day of Week model with the totals from the lowest level being propagated up the tree

## Implementation
This all works thanks to a `JS backend` script, a `relation`, and a `label`.

Here is the script:
```
// CONFIG
const FORMAT = 'YYYY-MM-DD HH:mm:ss';
// END CONFIG

const note = api.originEntity;
const parent = note.getParentNotes()[0];
let parentTitlePrefix = getTitlePrefix(parent.title);

const siblings = parent.getChildNotes();

let nowString = api.dayjs().format(FORMAT);

// get list of clock ins
const ins = siblings.filter(note => note.title.startsWith('IN - ')).map(note => note.title.replace('IN - ',''));

// get list of clock outs
const outs = siblings.filter(note => note.title.startsWith('OUT - ')).map(note => note.title.replace('OUT - ',''))

if (ins.length > outs.length) {
    // this new note must be a clock out
    note.title = `OUT - ${nowString}`;
    note.save();
    outs.push(nowString);
} else {
    // this new note must be a clock in
    note.title = `IN - ${nowString}`;
    note.save();
    ins.push(nowString);
}

// update parent note's title
if (ins.length == outs.length) {
    // you just clocked out. calculate total time
    
    let totalTimeInMins = 0;
    for (let i=0,size=outs.length; i<size; i++) {
        const inTime = api.dayjs(ins[i], FORMAT);
        const outTime = api.dayjs(outs[i], FORMAT);
        totalTimeInMins += api.dayjs(outTime).diff(inTime, 'minute', true);
    }
    
    const totalTimeInHours = round(totalTimeInMins/60);
    
    parent.title = `${parentTitlePrefix} (${totalTimeInHours} hrs)`;
    parent.save();
    
    // accumulate totals up the tree
    let grandparent = parent.getParentNotes()[0];
    while (grandparent.getLabel('sumTimesheet')) {
        sumUpTheTree(grandparent);
        grandparent = grandparent.getParentNotes()[0];
    }
} else {
    // you just clocked in. change title to ON THE CLOCK
    parent.title = `${parentTitlePrefix} (ON THE CLOCK)`;
    parent.save();
}

function sumUpTheTree(note) {
    const children = note.getChildNotes();

    if (children.some(note => note.title.indexOf('ON THE CLOCK') != -1)) {
        // you're on the clock right now, so don't total anything
        return;
    }

    const sumFunction = (sum, val) => sum + val;

    // get all the hour totals from each child's title and add them together
    let totalHours = children.filter(note => note.title.indexOf(' (') != -1).map(note => parseFloat(note.title.substring(note.title.indexOf('(')+1, note.title.indexOf(')')))).reduce(sumFunction);
    
    totalHours = round(totalHours);
    
    let noteTitlePrefix = getTitlePrefix(note.title);
    
    note.title = `${noteTitlePrefix} (${totalHours} hrs)`;
    note.save();
}

function round(num) {
    // rounds to 2 decimal places
    return Math.round(num * 100) / 100;
}

function getTitlePrefix(title) {
    // if there are already () in the title, get everything before them
    if (title.indexOf(' (') >= 0) {
        return title.substring(0, title.indexOf(' ('));
    }
    return title;
}
```

To create a timesheet, add this relation to the lowest level of your nested structure: `~runOnChildNoteCreation=@js` where `@js` is the link to the above `JS backend` script.

In Example 1 above, the following notes have this relation:
- Task 1
- Task 2 / Sub-task 1
- Task 2 / Sub-task 2
- Task 3 / Sub-task 1
- Task 3 / Sub-task 2

To sum up the totals from nested timesheets, add this label to the parent note: `#sumTimesheet`.

In Example 1 above, the following notes have this label:
- Project 1
- Task 1
- Task 2

Since Task 3 does not have the label `#sumTimesheet`, the totals of its Sub-task 1 and Sub-task 2 are not combined into the Task 3 title nor are they added to Project 1's total.
